# #Dataset Overview

*Number of observations (n): 2938

*Number of predictors (p) :21

n≥p+5⇒2938≥21+5   
so the data is suitable for multivariable regression

# #features explaination
###Demographic & Identifiers

*   Country
*   year: Year of observation (2000–2015)
*   Status: Developing or Developed country.

###Health & Mortality Indicators

*   Adult Mortality:Probability of dying between ages 15–60 per 1000 population.
*   infant deaths: Number of infant deaths per 1000 live births.
*   under-five deaths: Deaths of children under 5 per 1000 population.
*   Hepatitis B : Immunization coverage against the Hepatitis B.
*   Polio : Polio immunization rate.
*   Diphtheria: DPT vaccination coverage.
*   Measles: number of reported cases per 1000 population.
*   HIV/AIDS: Deaths per 1000 population.

##Lifestyle & Physical Health


*   Alcohol:The average annual alcohol consumption per individual.
*   BMI:Average Body Mass Index.  
*   thinness 1–19 years: Malnutrition indicator for youth.
*   thinness 5–9 years: Malnutrition indicator for children.

##Economic Indicators
*   GDP:The average income generated per person in a country
*   percentage expenditure:Health expenditure as a percentage of GDP.
*   Total expenditure:Government health spending (%).   
*  Income composition of resources:Composite income index; strong predictor of development.

##Education & Social Development

*  Schooling : Average years of schooling
*  Population : The total population size of a country



In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Load the dataset

In [ ]:
data = pd.read_csv('/content/Life Expectancy Data.csv')

In [ ]:
data

Country  Year      Status  Life expectancy   Adult Mortality  \
0     Afghanistan  2015  Developing         65.000000       263.000000   
1     Afghanistan  2014  Developing         59.900000       271.000000   
2     Afghanistan  2013  Developing         59.900000       268.000000   
3     Afghanistan  2012  Developing         59.500000       272.000000   
4     Afghanistan  2011  Developing         59.200000       275.000000   
...           ...   ...         ...               ...              ...   
2933     Zimbabwe  2004  Developing         44.300000       723.000000   
2934     Zimbabwe  2003  Developing         44.500000       715.000000   
2935     Zimbabwe  2002  Developing         44.800000        73.000000   
2936     Zimbabwe  2001  Developing         45.300000       686.000000   
2937     Zimbabwe  2000  Developing         46.000000       665.000000   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
0                62 0.010000               71.279624    65.000000      1154   
1                64 0.010000               73.523582    62.000000       492   
2                66 0.010000               73.219243    64.000000       430   
3                69 0.010000               78.184215    67.000000      2787   
4                71 0.010000                7.097109    68.000000      3013   
...             ...      ...                     ...          ...       ...   
2933             27 4.360000                0.000000    68.000000        31   
2934             26 4.060000                0.000000     7.000000       998   
2935             25 4.430000                0.000000    73.000000       304   
2936             25 1.720000                0.000000    76.000000       529   
2937             24 1.680000                0.000000    79.000000      1483   

      ...     Polio  Total expenditure  Diphtheria    HIV/AIDS        GDP  \
0     ...  6.000000           8.160000    65.000000   0.100000 584.259210   
1     ... 58.000000           8.180000    62.000000   0.100000 612.696514   
2     ... 62.000000           8.130000    64.000000   0.100000 631.744976   
3     ... 67.000000           8.520000    67.000000   0.100000 669.959000   
4     ... 68.000000           7.870000    68.000000   0.100000  63.537231   
...   ...       ...                ...          ...        ...        ...   
2933  ... 67.000000           7.130000    65.000000  33.600000 454.366654   
2934  ...  7.000000           6.520000    68.000000  36.700000 453.351155   
2935  ... 73.000000           6.530000    71.000000  39.800000  57.348340   
2936  ... 76.000000           6.160000    75.000000  42.100000 548.587312   
2937  ... 78.000000           7.100000    78.000000  43.500000 547.358878   

          Population   thinness  1-19 years   thinness 5-9 years  \
0    33736494.000000              17.200000            17.300000   
1      327582.000000              17.500000            17.500000   
2    31731688.000000              17.700000            17.700000   
3     3696958.000000              17.900000            18.000000   
4     2978599.000000              18.200000            18.200000   
...              ...                    ...                  ...   
2933 12777511.000000               9.400000             9.400000   
2934 12633897.000000               9.800000             9.900000   
2935   125525.000000               1.200000             1.300000   
2936 12366165.000000               1.600000             1.700000   
2937 12222251.000000              11.000000            11.200000   

      Income composition of resources  Schooling  
0                            0.479000  10.100000  
1                            0.476000  10.000000  
2                            0.470000   9.900000  
3                            0.463000   9.800000  
4                            0.454000   9.500000  
...                               ...        ...  
2933                         0.407000   9.200000  
2934                  

# Check the dimension of the dataset

In [ ]:
data.shape

(2938, 22)

2938 rows → number of observations (n)

22 columns → number of variables (features + target)



# Get information about the data

In [ ]:
# Display column names and data types
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

# Classification of Variables in the Dataset


In [ ]:
# Categorical variables
categorical_variables = ["Country", "Status"]

# Numerical variables
numerical_variables = data.columns.drop(categorical_variables)

numerical_variables

Index(['Year', 'Life expectancy ', 'Adult Mortality', 'infant deaths',
       'Alcohol', 'percentage expenditure', 'Hepatitis B', 'Measles ', ' BMI ',
       'under-five deaths ', 'Polio', 'Total expenditure', 'Diphtheria ',
       ' HIV/AIDS', 'GDP', 'Population', ' thinness  1-19 years',
       ' thinness 5-9 years', 'Income composition of resources', 'Schooling'],
      dtype='object')

In [ ]:
data["Country"].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Australia',
       'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan',
       'Bolivia (Plurinational State of)', 'Bosnia and Herzegovina',
       'Botswana', 'Brazil', 'Brunei Darussalam', 'Bulgaria',
       'Burkina Faso', 'Burundi', "Côte d'Ivoire", 'Cabo Verde',
       'Cambodia', 'Cameroon', 'Canada', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Cook Islands', 'Costa Rica', 'Croatia', 'Cuba', 'Cyprus',
       'Czechia', "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Denmark', 'Djibouti',
       'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Ethiopia', 'Fiji', 'Finland', 'France', 'Gabon', 'Gambia',
       'Georgia', 'Germany'

In [ ]:
data["Country"].unique().size

193

Country is a nominal categorical variable → have 193 unique categories, would create hundreds of dummy variables so we will drop this column instead of making one hot encoding and the country does not affect the life expectancy

Status is a binary categorical variable →Label Encoding

```
Developed → 1, Developing → 0
```






# Encoding for categorical features

In [ ]:
# Clean Status column first to avoid missing value problem in status after encoding
data["Status"] = data["Status"].str.strip()

# Encode Status
data["Status"] = data["Status"].map({"Developed": 1, "Developing": 0})

# Drop Country column
data.drop(columns=["Country"], inplace=True)



In [ ]:
data['Status']

0       0
1       0
2       0
3       0
4       0
       ..
2933    0
2934    0
2935    0
2936    0
2937    0
Name: Status, Length: 2938, dtype: int64

# Verify encoding

In [ ]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 21 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Year                             2938 non-null   int64  
 1   Status                           2938 non-null   int64  
 2   Life expectancy                  2928 non-null   float64
 3   Adult Mortality                  2928 non-null   float64
 4   infant deaths                    2938 non-null   int64  
 5   Alcohol                          2744 non-null   float64
 6   percentage expenditure           2938 non-null   float64
 7   Hepatitis B                      2385 non-null   float64
 8   Measles                          2938 non-null   int64  
 9    BMI                             2904 non-null   float64
 10  under-five deaths                2938 non-null   int64  
 11  Polio                            2919 non-null   float64
 12  Total expenditure   

# Count missing values for each column


In [ ]:
missing_count = data.isnull().sum()
missing_count

Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

# Drop the missing values in target variable ['Life expectancy']

In [ ]:
data = data.dropna(subset=['Life expectancy '])

#Immunization columns (Hepatitis B – Polio – Diphtheria)
 Missing values in immunization variables were imputed using group-wise median
 based on country development status to preserve realistic differences
 between developed and developing countries.


In [ ]:
# Group-wise median imputation for immunization variables
immunization_cols = ["Hepatitis B", "Polio", "Diphtheria "]

for col in immunization_cols:
    data[col] = data.groupby("Status")[col].transform(
        lambda x: x.fillna(x.median())
    )

/tmp/ipython-input-381129126.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[col] = data.groupby("Status")[col].transform(


#Social Variables (Schooling – Income composition)
Social variables are imputed using group-wise median by development status
 because education and income composition are strongly associated with a
 country's level of development.

In [ ]:
social_cols = ["Schooling", "Income composition of resources"]

for col in social_cols:
    data[col] = data.groupby("Status")[col].transform(
        lambda x: x.fillna(x.median())
    )


/tmp/ipython-input-3470988948.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[col] = data.groupby("Status")[col].transform(


#Economic variables
GDP and Population are imputed using group-wise median due to strong
 association with development status.

Total health expenditure is imputed using the global median.

In [ ]:
# GDP and Population are imputed using group-wise median
for col in ["GDP", "Population"]:
    data[col] = data.groupby("Status")[col].transform(
        lambda x: x.fillna(x.median())
    )

# Total health expenditure is imputed using the global median.
data["Total expenditure"] = data["Total expenditure"].fillna(
    data["Total expenditure"].median()
)


/tmp/ipython-input-3436466739.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[col] = data.groupby("Status")[col].transform(
/tmp/ipython-input-3436466739.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Total expenditure"] = data["Total expenditure"].fillna(


## Lifestyle-related variables are imputed using the median of each column.

In [ ]:
lifestyle_cols = ["Alcohol",' BMI '," thinness  1-19 years", " thinness 5-9 years"]

# Impute lifestyle variables using median
for col in lifestyle_cols:
    data.loc[:, col] = data[col].fillna(data[col].median())

#check that no missing values remain in the dataset

In [ ]:
data.isnull().sum()

Year                               0
Status                             0
Life expectancy                    0
Adult Mortality                    0
infant deaths                      0
Alcohol                            0
percentage expenditure             0
Hepatitis B                        0
Measles                            0
 BMI                               0
under-five deaths                  0
Polio                              0
Total expenditure                  0
Diphtheria                         0
 HIV/AIDS                          0
GDP                                0
Population                         0
 thinness  1-19 years              0
 thinness 5-9 years                0
Income composition of resources    0
Schooling                          0
dtype: int64

## Outlier Detection using IQR

In [ ]:
outlier_summary = {}  # Dictionary to store the number of outliers for each column

for col in numerical_variables:
    Q1 = data[col].quantile(0.25)  # 25th percentile
    Q3 = data[col].quantile(0.75)  # 75th percentile
    IQR = Q3 - Q1                  # Interquartile range (IQR)

    lower_bound = Q1 - 1.5 * IQR   # Lower bound
    upper_bound = Q3 + 1.5 * IQR   # Upper bound

    # Select rows with value is below lower_bound or above upper_bound
    outliers = data[(data[col] < lower_bound) | (data[col] > upper_bound)]

    outlier_summary[col] = outliers.shape[0]  # Store the count of outliers

# Convert the dictionary to a DataFrame
pd.DataFrame.from_dict(outlier_summary, orient="index", columns=["Outlier Count"])


Outlier Count
Year                                         0
Life expectancy                             10
Adult Mortality                             82
infant deaths                              315
Alcohol                                      3
percentage expenditure                     388
Hepatitis B                                320
Measles                                    542
 BMI                                         0
under-five deaths                          394
Polio                                      278
Total expenditure                           49
Diphtheria                                 297
 HIV/AIDS                                  542
GDP                                        459
Population                                 452
 thinness  1-19 years                      100
 thinness 5-9 years                         99
Income composition of resources            130
Schooling                                   59

#Outliers Capping
 ### Cap values below the lower bound to the lower bound
 ### Cap values above the upper bound to the upper bound

In [ ]:
for col in numerical_variables:
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR  # Lower bound for capping
    upper_bound = Q3 + 1.5 * IQR  # Upper bound for capping

    # Capping
    data[col] = np.where(data[col] < lower_bound, lower_bound, data[col])
    data[col] = np.where(data[col] > upper_bound, upper_bound, data[col])


/tmp/ipython-input-33854993.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[col] = np.where(data[col] < lower_bound, lower_bound, data[col])
/tmp/ipython-input-33854993.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[col] = np.where(data[col] > upper_bound, upper_bound, data[col])
/tmp/ipython-input-33854993.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the doc

#Check for remaining outliers after capping


In [ ]:
outlier_summary_after = {}

for col in numerical_variables:
    Q1 = data[col].quantile(0.25)
    Q3 = data[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = data[(data[col] < lower_bound) | (data[col] > upper_bound)]
    outlier_summary_after[col] = outliers.shape[0]

pd.DataFrame.from_dict(outlier_summary_after, orient="index", columns=["Outlier Count"])


Outlier Count
Year                                         0
Life expectancy                              0
Adult Mortality                              0
infant deaths                                0
Alcohol                                      0
percentage expenditure                       0
Hepatitis B                                  0
Measles                                      0
 BMI                                         0
under-five deaths                            0
Polio                                        0
Total expenditure                            0
Diphtheria                                   0
 HIV/AIDS                                    0
GDP                                          0
Population                                   0
 thinness  1-19 years                        0
 thinness 5-9 years                          0
Income composition of resources              0
Schooling                                    0

#Descriptive Statistics

In [ ]:
def compute_mean(x):
    return sum(x) / len(x)



In [ ]:
def compute_median(x):
    x_sorted = sorted(x)
    n = len(x_sorted)
    mid = n // 2
    if n % 2 == 0:
        return (x_sorted[mid - 1] + x_sorted[mid]) / 2
    else:
        return x_sorted[mid]


In [ ]:
def compute_variance(x):
    mean = compute_mean(x)
    return sum((xi - mean)**2 for xi in x) / (len(x) - 1)


In [ ]:
def compute_std(x):
    return compute_variance(x) ** 0.5


In [ ]:
def compute_mode(x):
    freq = {}
    for value in x:
        freq[value] = freq.get(value, 0) + 1
    max_freq = max(freq.values())
    modes = [k for k, v in freq.items() if v == max_freq]
    return modes[0]  # return first mode


In [ ]:
def min_value(x):
    m = x[0]
    for value in x:
        if value < m:
            m = value
    return m

def max_value(x):
    m = x[0]
    for value in x:
        if value > m:
            m = value
    return m


In [ ]:
pd.options.display.float_format = '{:.6f}'.format


#Measures of Dispersion

In [ ]:
dispersion = {}

for col in numerical_variables:
    values = data[col].dropna().tolist()
    dispersion[col] = {
        "Variance": compute_variance(values),
        "Standard Deviation": compute_std(values),
        "Min Value": min_value(values),
        "Max Value": max_value(values)
    }
dispersion_df = pd.DataFrame(dispersion).T
dispersion_df


Variance  Standard Deviation  \
Year                                        21.257260            4.610560   
Life expectancy                             90.251470            9.500077   
Adult Mortality                          13425.089722          115.866689   
infant deaths                              365.761305           19.124887   
Alcohol                                     15.356264            3.918707   
percentage expenditure                  152401.043890          390.385763   
Hepatitis B                                152.337973           12.342527   
Measles                                 126518.876409          355.694920   
 BMI                                       394.312678           19.857308   
under-five deaths                          635.825697           25.215584   
Polio                                      249.757813           15.803728   
Total expenditure                            5.229448            2.286799   
Diphtheria                                 249.155923           15.784674   
 HIV/AIDS                                    0.481784            0.694107   
GDP                                   20340615.267761         4510.057125   
Population                      14823760642443.921875      3850163.716317   
 thinness  1-19 years                       15.475885            3.933940   
 thinness 5-9 years                         15.814671            3.976766   
Income composition of resources              0.038280            0.195652   
Schooling                                    9.953422            3.154904   

                                  Min Value       Max Value  
Year                            2000.000000     2015.000000  
Life expectancy                   44.200000       89.000000  
Adult Mortality                    1.000000      459.000000  
infant deaths                      0.000000       55.000000  
Alcohol                            0.010000       16.838750  
percentage expenditure             0.000000     1099.254858  
Hepatitis B                       61.000000       99.000000  
Measles                            0.000000      905.625000  
 BMI                               1.000000       77.600000  
under-five deaths                  0.000000       70.000000  
Polio                             49.500000       99.000000  
Total expenditure                  0.370000       11.770000  
Diphtheria                        49.500000       99.000000  
 HIV/AIDS                          0.100000        1.850000  
GDP                                1.681350    12765.010932  
Population                        34.000000 10854761.125000  
 thinness  1-19 years              0.100000       15.350000  
 thinness 5-9 years                0.100000       15.600000  
Income composition of resources    0.089625        0.948000  
Schooling                          4.300000       20.300000

## Measures of Central Tendency


In [ ]:
central_tendency = {}

for col in numerical_variables:
    values = data[col].dropna().tolist()
    central_tendency[col] = {
        "Mean": compute_mean(values),
        "Median": compute_median(values),
        "Mode": compute_mode(values)
    }

central_df = pd.DataFrame(central_tendency).T
central_df


Mean         Median           Mode
Year                               2007.500000    2007.500000    2015.000000
Life expectancy                      69.233197      72.100000      73.000000
Adult Mortality                     162.087432     144.000000     459.000000
infant deaths                        13.681694       3.000000       0.000000
Alcohol                               4.558602       3.770000       0.010000
percentage expenditure              284.889600      65.611455       0.000000
Hepatitis B                          87.057719      91.000000      91.000000
Measles                             235.469518      17.000000       0.000000
 BMI                                 38.291291      43.350000      43.350000
under-five deaths                    18.027322       4.000000       0.000000
Polio                                85.337432      93.000000      99.000000
Total expenditure                     5.886021       5.750000       5.750000
Diphtheria                           85.282616      93.000000      99.000000
 HIV/AIDS                             0.537705       0.100000       0.100000
GDP                                3863.743924    1305.304687   12765.010932
Population                      3181500.861735 1412966.500000 1412966.500000
 thinness  1-19 years                 4.698805       3.300000      15.350000
 thinness 5-9 years                   4.718545       3.400000      15.600000
Income composition of resources       0.635381       0.673000       0.089625
Schooling                            12.094638      12.300000      11.700000

In [ ]:
data.describe()

Year      Status  Life expectancy   Adult Mortality  \
count 2928.000000 2928.000000       2928.000000      2928.000000   
mean  2007.500000    0.174863         69.233197       162.087432   
std      4.610560    0.379915          9.500077       115.866689   
min   2000.000000    0.000000         44.200000         1.000000   
25%   2003.750000    0.000000         63.100000        74.000000   
50%   2007.500000    0.000000         72.100000       144.000000   
75%   2011.250000    0.000000         75.700000       228.000000   
max   2015.000000    1.000000         89.000000       459.000000   

       infant deaths     Alcohol  percentage expenditure  Hepatitis B  \
count    2928.000000 2928.000000             2928.000000  2928.000000   
mean       13.681694    4.558602              284.889600    87.057719   
std        19.124887    3.918707              390.385763    12.342527   
min         0.000000    0.010000                0.000000    61.000000   
25%         0.000000    1.107500                4.853964    82.000000   
50%         3.000000    3.770000               65.611455    91.000000   
75%        22.000000    7.400000              442.614322    96.000000   
max        55.000000   16.838750             1099.254858    99.000000   

         Measles         BMI   ...       Polio  Total expenditure  \
count 2928.000000 2928.000000  ... 2928.000000        2928.000000   
mean   235.469518   38.291291  ...   85.337432           5.886021   
std    355.694920   19.857308  ...   15.803728           2.286799   
min      0.000000    1.000000  ...   49.500000           0.370000   
25%      0.000000   19.400000  ...   78.000000           4.370000   
50%     17.000000   43.350000  ...   93.000000           5.750000   
75%    362.250000   56.100000  ...   97.000000           7.330000   
max    905.625000   77.600000  ...   99.000000          11.770000   

       Diphtheria     HIV/AIDS          GDP      Population  \
count  2928.000000 2928.000000  2928.000000     2928.000000   
mean     85.282616    0.537705  3863.743924  3181500.861735   
std      15.784674    0.694107  4510.057125  3850163.716317   
min      49.500000    0.100000     1.681350       34.000000   
25%      78.000000    0.100000   578.797095   418120.500000   
50%      93.000000    0.100000  1305.304687  1412966.500000   
75%      97.000000    0.800000  5453.282630  4592776.750000   
max      99.000000    1.850000 12765.010932 10854761.125000   

        thinness  1-19 years   thinness 5-9 years  \
count            2928.000000          2928.000000   
mean                4.698805             4.718545   
std                 3.933940             3.976766   
min                 0.100000             0.100000   
25%                 1.600000             1.600000   
50%                 3.300000             3.400000   
75%                 7.100000             7.200000   
max                15.350000            15.600000   

       Income composition of resources   Schooling  
count                      2928.000000 2928.000000  
mean                          0.635381   12.094638  
std                           0.195652    3.154904  
min                           0.089625    4.300000  
25%                           0.504000   10.300000  
50%                           0.673000   12.300000  
75%                           0.780250   14.300000  
max                           0.948000   20.300000  

[8 rows x 21 columns]

#Split Data to Target & Features for regreesion model

In [ ]:
# Define target variable
y = data['Life expectancy ']

# Define feature matrix (drop target)
X = data.drop(columns=['Life expectancy '])


#Feature Standardization (Z-score)

In [ ]:
X_standardized = X.copy()


In [ ]:
for col in X_standardized.columns:
    values = X_standardized[col].values
    mean = compute_mean(values)
    std = compute_std(values)
    if std != 0: # avoid dividing by zero
     X_standardized[col] = [(x - mean) / std for x in values]
    else:
        # If the feature has zero variance, set standardized values to zero
        X_standardized[col] = 0

In [ ]:
X_standardized

Year    Status  Adult Mortality  infant deaths   Alcohol  \
0     1.626701 -0.460270         0.870937       2.160447 -1.160741   
1     1.409807 -0.460270         0.939982       2.160447 -1.160741   
2     1.192914 -0.460270         0.914090       2.160447 -1.160741   
3     0.976020 -0.460270         0.948612       2.160447 -1.160741   
4     0.759127 -0.460270         0.974504       2.160447 -1.160741   
...        ...       ...              ...            ...       ...   
2933 -0.759127 -0.460270         2.562536       0.696386 -0.050680   
2934 -0.976020 -0.460270         2.562536       0.644098 -0.127236   
2935 -1.192914 -0.460270        -0.768879       0.591810 -0.032817   
2936 -1.409807 -0.460270         2.562536       0.591810 -0.724372   
2937 -1.626701 -0.460270         2.562536       0.539522 -0.734580   

      percentage expenditure  Hepatitis B  Measles       BMI   \
0                  -0.547177    -1.787131  1.884074 -0.966460   
1                  -0.541429    -2.030194  0.721209 -0.991640   
2                  -0.542208    -1.868152  0.546903 -1.016819   
3                  -0.529490    -1.625090  1.884074 -1.041999   
4                  -0.711585    -1.544069  1.884074 -1.062143   
...                      ...          ...       ...       ...   
2933               -0.729764    -1.544069 -0.574845 -0.563586   
2934               -0.729764    -2.111214  1.884074 -0.583729   
2935               -0.729764    -1.138966  0.192666 -0.603873   
2936               -0.729764    -0.895904  0.825231 -0.624017   
2937               -0.729764    -0.652842  1.884074 -0.644160   

      under-five deaths      Polio  Total expenditure  Diphtheria    HIV/AIDS  \
0               2.061133 -2.267657           0.994394    -1.284956  -0.630602   
1               2.061133 -1.729809           1.003140    -1.475014  -0.630602   
2               2.061133 -1.476704           0.981275    -1.348309  -0.630602   
3               2.061133 -1.160323           1.151819    -1.158251  -0.630602   
4               2.061133 -1.097047           0.867579    -1.094899  -0.630602   
...                  ...       ...                ...          ...        ...   
2933            0.950709 -1.160323           0.543983    -1.284956   1.890624   
2934            0.911051 -2.267657           0.277234    -1.094899   1.890624   
2935            0.871393 -0.780666           0.281607    -0.904841   1.890624   
2936            0.831735 -0.590837           0.119809    -0.651430   1.890624   
2937            0.831735 -0.464285           0.530864    -0.461373   1.890624   

           GDP  Population   thinness  1-19 years   thinness 5-9 years  \
0    -0.727149    1.992970               2.707513             2.736257   
1    -0.720844   -0.741246               2.707513             2.736257   
2    -0.716620    1.992970               2.707513             2.736257   
3    -0.708147    0.133879               2.707513             2.736257   
4    -0.842607   -0.052700               2.707513             2.736257   
...        ...         ...                    ...                  ...   
2933 -0.755950    1.992970               1.195035             1.177201   
2934 -0.756175    1.992970               1.296714             1.302932   
2935 -0.843979   -0.793726              -0.889389            -0.859629   
2936 -0.735059    1.992970              -0.787710            -0.759045   
2937 -0.735331    1.992970               1.601752             1.629830   

      Income composition of resources  Schooling  
0                           -0.799278  -0.632234  
1                           -0.814611  -0.663931  
2                           -0.845278  -0.695627  
3                           -0.881056  -0.727324  
4                           -0.927056  -0.822414  
...                               ...        ...  
2933                        -1.167278  -0.917504  
2934                        -1.111056  -0.822414  
2935                        -1.065056  -0.663931  
2936            

#Save cleaned data

In [ ]:
import os

# Create the directory if it doesn't exist
os.makedirs('Dataset', exist_ok=True)

data.to_csv('Dataset/Life Expectancy Data Cleaned.csv', index=False)